File: src/first_draft/bfs.py
import sys
import Queue
import ramcloud

GRAPH_TABLE_NAME = "graph"
COORDINATOR_LOCATION = "fast+udp:host=127.0.0.1,port=12246"

def main():
    # get connected to ramcloud
    rc = ramcloud.RAMCloud()
    rc.connect(COORDINATOR_LOCATION)

    graph_tableid = rc.get_table_id(GRAPH_TABLE_NAME)

    # run breadth first search on the graph    
    source = sys.argv[1]
    queue = Queue.Queue()
    queue.put(source)
    shortest_path = {}
    shortest_path[source] = (source, 0)
    while not queue.empty():
        u = queue.get()
        path_to_u, dist_to_u = shortest_path[u]
        
        try:
            value, version = rc.read(graph_tableid, u)
            for v in value.split(' '):
                if v != "-1":
                    try:
                        shortest_path[v]
                    except:
                        queue.put(v)
                        shortest_path[v] = (path_to_u + "->" + v, dist_to_u + 1)
                else:
                    rc.write(graph_tableid, u, value[0:-2] + path_to_u + ' ' + str(dist_to_u))
        except:
            # this node is a leaf                   
            rc.write(graph_tableid, u, path_to_u + ' ' + str(dist_to_u))

if __name__ == '__main__':
    main()

File: src/first_draft/bfs.cc
#include <iostream>
#include <sstream>
#include <queue>
#include <set>
#include "RamCloud.h"

#define GRAPH_TABLE_NAME "graph"
#define COORDINATOR_LOCATION "infrc:host=192.168.1.101,port=12246"

using namespace RAMCloud;

int main(int argc, char* argv[]) { 
  uint64_t graph_tableid;

  RamCloud client(COORDINATOR_LOCATION);
 
  graph_tableid = client.getTableId(GRAPH_TABLE_NAME);

  std::queue<pair<string,int>> queue;
  std::set<string> seen_list;

  Buffer buffer;
  string u(argv[1]);
  uint32_t dist_to_u = 0;
  queue.push(pair<string, int>(u, dist_to_u)); 
  seen_list.insert(u);
  while( !queue.empty() ) {
    u = queue.front().first;
    dist_to_u = queue.front().second;
    std::stringstream ss;
    ss << dist_to_u;
    string dist_to_u_str = ss.str();
    queue.pop();

    try {
      client.read(graph_tableid, u.c_str(), u.length(), &buffer);
      
      string edge_list(static_cast<const char*>(buffer.getRange(0, buffer.getTotalLength())), buffer.getTotalLength());
    
      std::istringstream iss(edge_list);
    
      while(iss) {
        string v;
        iss >> v;
        if( v != "-1" ) {
          //std::cout << v << "\n";
          if( seen_list.count(v) == 0 ) {
            queue.push(pair<string, int>(v, dist_to_u + 1));
            seen_list.insert(v);
          }
        } else {
          //std::cout << "writing: " << dist_to_u_str << "\n";
          client.write(graph_tableid, u.c_str(), u.length(), edge_list.append(" " + dist_to_u_str).c_str());
          break;
        }
      }
    } catch (RAMCloud::ClientException& e) {
      //fprintf(stderr, "RAMCloud exception: %s\n", e.str().c_str());
      client.write(graph_tableid, u.c_str(), u.length(), dist_to_u_str.c_str());
    } catch (RAMCloud::Exception& e) {
      //fprintf(stderr, "RAMCloud exception: %s\n", e.str().c_str());
      client.write(graph_tableid, u.c_str(), u.length(), dist_to_u_str.c_str());
    }
  }

  // finally, drop the table
  //client.dropTable(GRAPH_TABLE_NAME);

  return 0;
}
